core/src/main/scala/org/http4s/syntax/KleisliSyntax.scala

```diff
@@ -30,6 +30,9 @@ trait KleisliSyntaxBinCompat1 {
 final class KleisliResponseOps[F[_]: Functor, A](self: Kleisli[OptionT[F, ?], A, Response[F]]) {
   def orNotFound: Kleisli[F, A, Response[F]] =
     Kleisli(a => self.run(a).getOrElse(Response.notFound))
+
+  def orNotFound(notFoundHandler: A => Response[F]): Kleisli[F, A, Response[F]] =
+    Kleisli(a => self.run(a).getOrElse(notFoundHandler(a)))
```

Comment:
I don't think this use case (a custom fall through handler) is a common enough use case to warrant special syntax. I can also think of a few other ways to implement similar functionality - instead of `A => Response[F]` we can use `Response[F]`, `F[Response[F]]`, `A => F[Response[F]]`, `HttpApp[F]`... All of those potential functions are equally valid, and would have their uses - I don't think the one you've chosen is any more common than the others.

If you insist, I'd like you to at least choose another name. Overloading should be avoided in general, and `orNotFound` is misleading here, since it has no relation to `NotFound`.

Reply:
It would be nice to be able to do:

```
JettyBuilder[IO]
      .bindAny()
      .mountServiceHttpApp(
        HttpRoutes
          .of[IO](Map.empty)
          .orNotFound(myCustomNotFoundHandler),
        "/"
      )
      .mountServiceHttpApp(
        HttpRoutes
          .of[IO] {
            case GET -> Root / "hello" => Ok("hello")
          }
          .orNotFound(myCustomNotFoundHandler),
        "/prefix1"
      )
```
So that you can see in one place how all `non-matched endpoints` are handled and that they are no leaks. The method name could be `.orHandleNonMatch`. I agree with all your other points.

Reply:
> So that you can see in one place how all non-matched endpoints are handled and that there are no leaks.

You don't need a (library) syntax for that - it works just as well with a custom function:
```scala
def orMyHandler(x: HttpRoutes[F]): HttpApp[F] =
  Kleisli { req => x.run(req).getOrElse(myCustomNotFoundHandler(req)) }

//...
.mountServiceHttpApp(
  orMyHandler(httpRoutes0),
  "/"
)
// ...
```

As an aside, I've never seen someone mount a lot of services onto one builder like this. I'd rewrite your example like this instead:

```scala
JettyBuilder[IO]
  .bindAny()
  .mountServiceHttpApp(
    myOrHandler(
      Router(
        "" -> httpRoutes0,
        "prefix1" -> httpRoutes1,
        ...
        "prefixN" -> httpRoutesN
      )
    )
  )
```

Reply:
Thank you, this is exactly what I have been looking for !

More generally, do you think there is a need for a `JettyBuilder.mountServiceHttpApp` ? I would say yes because as `.mountServiceHttpApp` expects a `total` function then it is clearer to see how `NotFound` for instance are handled. Also it would make sense to unify the APIs between `Jetty` and `Blaze`. What do you think ?

Reply:
Added clarity is not as important here; AFAICT, people tend to separate their routing logic from the server itself (http4s is great in that), so whoever wants to be sure they have matched all the possible options is probably already using `HttpApp` anyway, and just rewraps that as `HttpRoutes`. I think `mountServiceHttpApp` (or, rather, `mountHttpApp`) will be useful chiefly because it would allow us to avoid that annoying rewrapping.

> Also it would make sense to unify the APIs between Jetty and Blaze.

Not really. As it was mentioned here, Jetty and Blaze builders have fundamentally different capabilities for mounting routes, so I don't see an attempt to unify them leading to anything other than extra complexity and confusion.

Reply:
> I think mountServiceHttpApp (or, rather, mountHttpApp) will be useful chiefly because it would allow us to avoid that annoying rewrapping.

Useless rewrapping on the library-user side is always good, I will simplify this PR and follow your advices